tests/Aql/AqlValueMemoryLayoutTest.cpp

```diff
@@ -0,0 +1,419 @@
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Jan Steemann
+////////////////////////////////////////////////////////////////////////////////
+
+#include "gtest/gtest.h"
+
+#include "Aql/AqlValue.h"
+#include "Basics/Endian.h"
+
+#include <velocypack/Builder.h>
+#include <velocypack/Slice.h>
+#include <velocypack/velocypack-aliases.h>
+
+using namespace arangodb;
+using namespace arangodb::aql;
+
+namespace {
+    
+constexpr uint8_t UNINITIALIZED = 0xa5;
+
+void runChecksForNumber(AqlValue const& value, uint8_t const* expected) {
+  ASSERT_FALSE(value.requiresDestruction());
+  ASSERT_FALSE(value.isEmpty());
+  ASSERT_FALSE(value.isPointer());
+  ASSERT_FALSE(value.isManagedDocument());
+  ASSERT_FALSE(value.isRange());
+  ASSERT_FALSE(value.isNone());
+  ASSERT_FALSE(value.isNull(false));
+  ASSERT_FALSE(value.isNull(true));
+  ASSERT_FALSE(value.isBoolean());
+  ASSERT_TRUE(value.isNumber());
+  ASSERT_FALSE(value.isString());
+  ASSERT_FALSE(value.isObject());
+  ASSERT_FALSE(value.isArray());
+
+  if constexpr (arangodb::basics::isLittleEndian()) {
+    uint8_t const* data = reinterpret_cast<uint8_t const*>(&value);
+
+    for (uint8_t i = 0; i < sizeof(AqlValue); ++i) {
+      // note: in all tests, the value UNINITIALIZED (0xa5) has a special 
+      // meaning in a byte and will lead to it not being compared
+      if (expected[i] == UNINITIALIZED) {
+        continue;
+      }
+      ASSERT_EQ(data[i], expected[i]);
+    }
+  }
+}
+
+void runChecksForUInt64(uint64_t value, uint8_t const* expected) {
+  struct alignas(16) AqlValueMemory {
+    AqlValueMemory() {
+      // poison memory with some garbage values
+      memset(&buffer[0], 0x99, sizeof(buffer));
+    }
+    uint8_t buffer[16];
+  
+    static_assert(sizeof(buffer) == 16, "invalid size of AqlValueMemory buffer");
+  };
+  static_assert(sizeof(AqlValueMemory) == 16, "invalid size of AqlValueMemory");
+
+  // poison some memory with 0x99 values
+  AqlValueMemory memory;
+  void* p = reinterpret_cast<void*>(&memory);
+
+  // put uint64_t value into an AqlValue directly using the AqlValueHintUInt ctor.
+  // note: we are using placement new here, so that the AqlValue uses the
+  // poisoned memory region
+  new (p) AqlValue((AqlValueHintUInt(value)));
+  // although we are using placement new here, we don't need to call the
+  // destructor here, as the AqlValue with the payloads we use here won't
+  // do anything it is destructor
+
+  AqlValue& aqlValue = *reinterpret_cast<AqlValue*>(p);
+  runChecksForNumber(aqlValue, expected);
+
+  ASSERT_EQ(value != 0, aqlValue.toBoolean());
```

Comment:
```suggestion
  EXPECT_EQ(value != 0, aqlValue.toBoolean());
```
I prefer to use EXPECT instead of ASSERT for assertions that are not a precondition for following code, because that way the test does not abort with the first failing EXPECT and we get more information. As far as I see all ASSERTs in this test file could be converted to EXPECTs.
I only use ASSERT in cases where we definitely want to abort, because otherwise we would run into issues. E.g.:
```
  ASSERT_NE(nullptr, x);
  EXPECT_EQ(42, *x);
```

Reply:
Good idea, done that!